e2e/utils/test_cases.js

```diff
@@ -0,0 +1,163 @@
+
+            if (acc[key]) {
+                acc[key].push(item);
+            } else {
+                acc[key] = [item];
+            }
+
+            return acc;
+        }, {});
+}
+
+function saveToEndpoint(url, data) {
+    return axios({
+        method: 'POST',
+        url,
+        headers: {
+            'Content-Type': 'application/json; charset=utf-8',
+            Authorization: process.env.TM4J_API_KEY,
+        },
+        data,
+    }).catch((error) => {
+        console.log('Something went wrong:', error.response.data.message); // eslint-disable-line no-console
+        return error.response.data;
+    });
+}
+
+async function createTestCycle(startDate, endDate) {
+    const {
+        BRANCH,
+        BUILD_ID,
+        JIRA_PROJECT_KEY,
+        TM4J_FOLDER_ID,
+    } = process.env;
+
+    const testCycle = {
+        projectKey: JIRA_PROJECT_KEY,
+        name: `${BUILD_ID}-${BRANCH}`,
+        description: `Cypress automated test with ${BRANCH}`,
+        plannedStartDate: startDate,
+        plannedEndDate: endDate,
+        statusName: 'Done',
+        folderId: TM4J_FOLDER_ID,
+    };
+
+    const response = await saveToEndpoint('https://api.adaptavist.io/tm4j/v2/testcycles', testCycle);
+    return response.data;
+}
+
+async function createTestExecutions(report, testCycle) {
+    const {
+        BROWSER,
+        JIRA_PROJECT_KEY,
+    } = process.env;
+
+    const testCases = getTM4JTestCases(report);
+    const startDate = new Date(report.stats.start);
+    const startTime = startDate.getTime();
+
+    const promises = [];
+    Object.entries(testCases).forEach(([key, steps], index) => {
+        const testScriptResults = steps.
+            sort((a, b) => a.title.localeCompare(b.title)).
+            map((item) => {
+                return {
+                    statusName: status[item.state],
+                    actualEndDate: new Date(startTime + item.incrementalDuration).toISOString(),
+                    actualResult: 'Cypress automated test completed',
+                };
+            });
+
+        const stateResult = getStepStateResult(steps);
+
+        const testExecution = {
+            projectKey: JIRA_PROJECT_KEY,
+            testCaseKey: key,
+            testCycleKey: testCycle.key,
+            statusName: stateResult.passed && stateResult.passed === steps.length ? 'Pass' : 'Fail',
+            testScriptResults,
+            environmentName: environment[BROWSER] || 'Chrome',
+            actualEndDate: testScriptResults[testScriptResults.length - 1].actualEndDate,
+            executionTime: steps.reduce((acc, prev) => {
+                acc += prev.duration; // eslint-disable-line no-param-reassign
+                return acc;
+            }, 0),
+            comment: `Cypress automated test - ${getStepStateSummary(steps)}`,
+        };
+
+        // Temporarily log to verify cases that were being saved.
+        console.log(index, key); // eslint-disable-line no-console
```

Comment:
Intentionally added so I can verify manually the test cases being filtered and saved. Will probably remove in the future but is important to be here currently.